src/CloudFoundry.CloudController.Test.Integration/UploadAppTest.cs

```diff
@@ -0,0 +1,322 @@
+            File.WriteAllText(Path.Combine(tempAppPath, "content-unique-clone.txt"), content1);
+            File.WriteAllText(Path.Combine(tempAppPath, "content-unique2.txt"), content2);
+            File.WriteAllText(Path.Combine(tempAppPath, "zero.txt"), "");
+            File.WriteAllText(Path.Combine(tempAppPath, "zero2.txt"), "");
+
+            CreateAppResponse app = client.Apps.CreateApp(apprequest).Result;
+
+            Guid appGuid = app.EntityMetadata.Guid;
+
+            client.Apps.Push(appGuid, tempAppPath, false).Wait();
+
+            var zipFromServerPath = Path.GetTempFileName();
+            DownloadAppZip(appGuid.ToString(), zipFromServerPath);
+            if (new FileInfo(zipFromServerPath).Length == 0)
+            {
+                Assert.Inconclusive("API endpoint doesn't support file downloads");
+            }
+
+            using (var fs = File.OpenRead(zipFromServerPath))
+            {
+                using (var arch = new ZipArchive(fs, ZipArchiveMode.Read, true))
+                {
+                    Assert.IsTrue(arch.Entries.Count() == 7);
+
+                    Assert.IsTrue(arch.Entries.First(a => a.Name == "content.txt").Length == staticContent.Length);
+                    Assert.IsTrue(arch.Entries.First(a => a.Name == "zero.txt").Length == 0);
+
+                    var contentSStream = new StreamReader(arch.Entries.First(a => a.Name == "content-Unique.txt").Open());
+                    Assert.IsTrue(contentSStream.ReadToEnd() == content1);
+                }
+            }
+
+            client.Apps.DeleteApp(appGuid).Wait();
+            Directory.Delete(tempAppPath, true);
+        }
+
+        [TestMethod]
+        public void UploadTwiceTest()
+        {
+            var staticContent = "alabala";
+            var contentPath = Path.Combine(tempAppPath, "content.zip");
+
+            using (var fs = File.OpenWrite(contentPath))
+            {
+                using (var za = new ZipArchive(fs, ZipArchiveMode.Create, true))
+                {
+                    var ze = za.CreateEntry("content.txt", CompressionLevel.NoCompression);
+                    using (var sw = new StreamWriter(ze.Open()))
+                    {
+                        sw.Write(staticContent);
+                    }
+                }
+            }
+
+
+            CreateAppResponse app = client.Apps.CreateApp(apprequest).Result;
+            Guid appGuid = app.EntityMetadata.Guid;
+
+            using (var fs = File.OpenRead(contentPath))
+            {
+                client.Apps.UploadBits(appGuid, fs, new List<FileFingerprint>() {  }).Wait();
+            }
+
+            using (var fs = File.OpenRead(contentPath))
+            {
+                client.Apps.UploadBits(appGuid, fs, new List<FileFingerprint>() { }).Wait();
+            }
+
+            var zipFromServerPath = Path.GetTempFileName();
+            DownloadAppZip(appGuid.ToString(), zipFromServerPath);
```

Comment:
Could we verify the files uploaded by having the app start and report with signatures of its files?